// @gandlaf21/blind-signature/dist/lib/es5/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.hexToPoint = exports.pointToHex = exports.randomBytes = exports.randomPrivateKey = exports.SigningAuthority = exports.BlindedMessage = void 0;
var BlindedMessage_js_1 = require("./model/BlindedMessage.js");
Object.defineProperty(exports, "BlindedMessage", { enumerable: true, get: function () { return BlindedMessage_js_1.BlindedMessage; } });
var SigningAuthority_js_1 = require("./model/SigningAuthority.js");
Object.defineProperty(exports, "SigningAuthority", { enumerable: true, get: function () { return SigningAuthority_js_1.SigningAuthority; } });
var crypto_js_1 = require("./util/crypto.js");
Object.defineProperty(exports, "pointToHex", { enumerable: true, get: function () { return crypto_js_1.pointToHex; } });
Object.defineProperty(exports, "hexToPoint", { enumerable: true, get: function () { return crypto_js_1.hexToPoint; } });
var secp256k1_1 = require("@noble/secp256k1");
var randomPrivateKey = secp256k1_1.utils.randomPrivateKey;
exports.randomPrivateKey = randomPrivateKey;
var randomBytes = secp256k1_1.utils.randomBytes;
exports.randomBytes = randomBytes;
